pkg/loop/internal/capabilities_registry.go

```diff
@@ -0,0 +1,318 @@
+	return client, nil
+
+}
+
+func (cr *capabilitiesRegistryClient) GetConsensus(ctx context.Context, ID string) (capabilities.ConsensusCapability, error) {
+	req := &pb.GetConsensusRequest{
+		Id: ID,
+	}
+
+	res, err := cr.grpc.GetConsensus(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := cr.Dial(res.CapabilityID)
+	if err != nil {
+		return nil, ErrConnDial{Name: "GetAction", ID: res.CapabilityID, Err: err}
+	}
+	client := NewConsensusCapabilityClient(cr.BrokerExt, conn)
+	return client, nil
+}
+
+func (cr *capabilitiesRegistryClient) GetTarget(ctx context.Context, ID string) (capabilities.TargetCapability, error) {
+	req := &pb.GetTargetRequest{
+		Id: ID,
+	}
+
+	res, err := cr.grpc.GetTarget(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := cr.Dial(res.CapabilityID)
+	if err != nil {
+		return nil, ErrConnDial{Name: "GetAction", ID: res.CapabilityID, Err: err}
+	}
+	client := NewTargetCapabilityClient(cr.BrokerExt, conn)
+	return client, nil
+}
+
+func (cr *capabilitiesRegistryClient) List(ctx context.Context) ([]capabilities.BaseCapability, error) {
+	res, err := cr.grpc.List(ctx, &emptypb.Empty{})
+	if err != nil {
+		return nil, err
+	}
+
+	var clients []capabilities.BaseCapability
+	for _, id := range res.CapabilityID {
+		conn, err := cr.Dial(id)
+		if err != nil {
+			return nil, ErrConnDial{Name: "List", ID: id, Err: err}
+		}
+		client := newBaseCapabilityClient(cr.BrokerExt, conn)
+		clients = append(clients, client)
+	}
+
+	return clients, nil
+}
+
+func (cr *capabilitiesRegistryClient) Add(ctx context.Context, c capabilities.BaseCapability) error {
+	var cRes Resource
+	id, cRes, err := cr.ServeNew("someCapability", func(s *grpc.Server) {
+		pb.RegisterBaseCapabilityServer(s, &baseCapabilityServer{impl: c})
+	})
+
+	if err != nil {
+		return err
+	}
+
+	info, err := c.Info(ctx)
+	if err != nil {
+		cRes.Close()
+		return err
+	}
+
+	req := &pb.AddRequest{
+		CapabilityID: id,
+		Type:         pb.ExecuteAPIType(info.CapabilityType), // TODO: probably not right, figure out what this is
+	}
+
+	_, err = cr.grpc.Add(ctx, req)
+	if err != nil {
+		cRes.Close()
+		return err
+	}
+	return nil
+}
+
+func NewCapabilitiesRegistryClient(cc grpc.ClientConnInterface) *capabilitiesRegistryClient {
+	return &capabilitiesRegistryClient{grpc: pb.NewCapabilitiesRegistryClient(cc)}
+}
+
+var _ pb.CapabilitiesRegistryServer = (*capabilitiesRegistryServer)(nil)
+
+type capabilitiesRegistryServer struct {
+	pb.UnimplementedCapabilitiesRegistryServer
+	*BrokerExt
+	impl types.CapabilitiesRegistry
+}
+
+func (c *capabilitiesRegistryServer) Get(ctx context.Context, request *pb.GetRequest) (*pb.GetReply, error) {
+	capability, err := c.impl.Get(ctx, request.Id)
+	if err != nil {
+		return nil, err
+	}
+
+	var cRes Resource
+	id, cRes, err := c.ServeNew("someCapability", func(s *grpc.Server) {
+		pb.RegisterBaseCapabilityServer(s, &baseCapabilityServer{impl: capability})
```

Comment:
When you get a capability, you need to wrap it in the correct underlying client type.